internal/lib/stats/stats_server_linux.go

```diff
@@ -119,6 +149,101 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
 	})
 }
 
+// metricsForPodSandbox is an internal, non-locking version of MetricsForPodSandbox
+// that returns (and occasionally gathers) the metrics for the given sandbox.
+// Note: caller must hold the lock on the StatsServer
+func (ss *StatsServer) metricsForPodSandbox(sb *sandbox.Sandbox) *SandboxMetrics {
+	if ss.collectionPeriod == 0 {
+		return ss.updatePodSandboxMetrics(sb)
+	}
+	if sboxMetrics, ok := ss.sboxMetrics[sb.ID()]; ok {
+		return sboxMetrics
+	}
+	// Cache miss, try again.
+	return ss.updatePodSandboxMetrics(sb)
+}
+
+// updatePodSandboxMetrics updates the sandbox metrics for the given sandbox.
+// If the sandbox is not found, it creates a new entry in the map.
+// Note: caller must hold the lock on the StatsServer.
+func (ss *StatsServer) updatePodSandboxMetrics(sb *sandbox.Sandbox) *SandboxMetrics {
+	if sb == nil {
+		return nil
+	}
+	sm, exists := ss.sboxMetrics[sb.ID()]
+	if !exists {
+		sm = NewSandboxMetrics(sb)
+	}
+	// Network metrics are collected at the pod level.
+	if slices.Contains(ss.Config().IncludedPodMetrics, "network") {
+		podMetrics := ss.GenerateNetworkMetrics(sb)
+		sm.metric.Metrics = podMetrics
+	}
+
+	containersList := sb.Containers().List()
+	containerMetrics := make([]*types.ContainerMetrics, 0, len(containersList))
+	for _, c := range containersList {
+		// Skip if the container is stopped.
+		if c.StateNoLock().Status == oci.ContainerStateStopped {
+			continue
+		}
+		cMetrics := ss.GenerateSandboxContainerMetrics(sb, c, sm)
+		containerMetrics = append(containerMetrics, cMetrics)
+	}
+	sm.metric.ContainerMetrics = containerMetrics
+	ss.sboxMetrics[sb.ID()] = sm
+	return sm
+}
+
+// GenerateSandboxContainerMetrics generates a list of metrics for the specified sandbox
+// containers by collecting metrics from the cgroup based on the included pod metrics,
+// except for network metrics, which are collected at the pod level.
+func (ss *StatsServer) GenerateSandboxContainerMetrics(sb *sandbox.Sandbox, c *oci.Container, sm *SandboxMetrics) *types.ContainerMetrics {
+	cgstats, err := ss.Runtime().ContainerStats(context.TODO(), c, sb.CgroupParent())
+	if err != nil || cgstats == nil {
+		logrus.Errorf("Error getting sandbox stats %s: %v", sb.ID(), err)
+		return nil
```

Comment:
We should pass around the context here and use our internal `log` package everywhere.